### PR TITLE
Add thandle initialize move

### PR DESCRIPTION
**Pull request by dcristoloveanu**
Add thandle initialize move

### CODE REVIEW

File: devdoc/thandle_requirements.md [hunk 1]
```diff
@@ -223,4 +223,20 @@ void THANDLE_MOVE(T)(THANDLE(T) * t1, THANDLE(T) * t2)
...
+void THANDLE_INITIALIZE_MOVE(T)(THANDLE(T) * t1, THANDLE(T) * t2)
+```
+
+`THANDLE_INITIALIZE_MOVE` moves `*t2` under `*t1` and NULLs *t2, assuming that `t1` is not initialized. 
+
+**SRS_THANDLE_01_001: [** If `t1` is `NULL` then `THANDLE_INITIALIZE_MOVE` shall return. **]**
+
+**SRS_THANDLE_01_002: [** If `t2` is `NULL` then `THANDLE_INITIALIZE_MOVE` shall return. **]**
+
+**SRS_THANDLE_01_003: [** If `*t2` is `NULL` then `THANDLE_INITIALIZE_MOVE` shall `THANDLE_DEC_REF` `*t1`, set `*t1` to `NULL` and return. **]**
```
Comment (mattdurak) [inline devdoc/thandle_requirements.md:240]: >`THANDLE_DEC_REF` `*t1`, [](start = 82, length = 24)

shouldn't dec_ref

File: inc/azure_c_util/thandle.h [hunk 2]
```diff
@@ -411,6 +416,34 @@ void THANDLE_MOVE(T)(THANDLE(T) * t1, THANDLE(T) * t2 )
...
+        LogError("invalid argument THANDLE(" MU_TOSTRING(T) ") * t1=%p, THANDLE(" MU_TOSTRING(T) ") t2=%p", t1, t2 );                                               \
+    }                                                                                                                                                               \
+    else                                                                                                                                                            \
+    {                                                                                                                                                               \
+        if (*t2 == NULL)                                                                                                                                            \
+        {                                                                                                                                                           \
+            /*Codes_SRS_THANDLE_01_003: [ If *t2 is NULL then THANDLE_INITIALIZE_MOVE shall THANDLE_DEC_REF *t1, set *t1 to NULL and return. ]*/                    \
```
Comment (amarsa-msft) [inline inc/azure_c_util/thandle.h]: where's teh DEC_REF?

File: devdoc/thandle_requirements.md (hunk 1 above)
Comment (dcristoloveanu) [reply devdoc/thandle_requirements.md]: Spec fixed, we should not dec ref anything.

---
In reply to: [470882740](https://github.com/Azure/azure-c-util/pull/60#discussion_r470882740) [](ancestors = 470882740)

File: inc/azure_c_util/thandle.h [hunk 3]
```diff
@@ -411,6 +416,34 @@ void THANDLE_MOVE(T)(THANDLE(T) * t1, THANDLE(T) * t2 )
...
+            *(T const**)t1 = NULL;                                                                                                                                  \
+        }                                                                                                                                                           \
+        else                                                                                                                                                        \
+        {                                                                                                                                                           \
+            /*Codes_SRS_THANDLE_01_004: [ If *t2 is not NULL then THANDLE_INITIALIZE_MOVE shall set *t1 to *t2, set *t2 to NULL and return. ]*/                     \
```
Comment (amarsa-msft) [inline inc/azure_c_util/thandle.h:441]: why is there no DEC_REF here?

File: inc/azure_c_util/thandle.h (hunk 2 above)
Comment (dcristoloveanu) [reply inc/azure_c_util/thandle.h]: Spec fixed, we should not dec ref anything.

---
In reply to: [470884242](https://github.com/Azure/azure-c-util/pull/60#discussion_r470884242) [](ancestors = 470884242)

File: inc/azure_c_util/thandle.h (hunk 3 above)
Comment (dcristoloveanu) [reply inc/azure_c_util/thandle.h:441]: It is a move, the instance moves from t2 to t1. No ref count touches are needed.

---
In reply to: [470884539](https://github.com/Azure/azure-c-util/pull/60#discussion_r470884539) [](ancestors = 470884539)

**Pull request merged**